config/config_test.go

```diff
@@ -13,7 +13,7 @@ func TestRemoteDefault(t *testing.T) {
 	cfg := NewFrom(Values{
 		Git: map[string][]string{
 			"branch.unused.remote":     []string{"a"},
-			"branch.unused.pushRemote": []string{"b"},
```

Comment:
One thing that we could do to harden the use of `config.Values` throughout these tests against adding _new_ non-URL keys containing uppercase characters is to warn about it during construction, something like:

```diff
diff --git a/config/config.go b/config/config.go
index 6e38c595..112bf972 100644
--- a/config/config.go
+++ b/config/config.go
@@ -11,6 +11,7 @@ import (
 	"strings"
 	"sync"
 	"time"
+	"unicode"
 
 	"github.com/git-lfs/git-lfs/fs"
 	"github.com/git-lfs/git-lfs/git"
@@ -155,6 +156,13 @@ func NewFrom(v Values) *Configuration {
 			}
 
 			for key, values := range v.Git {
+				parts := strings.Split(key, ".")
+				isUrl := len(parts) >= 3
+				hasUpper := strings.IndexFunc(key, unicode.IsUpper) > -1
+
+				if !isUrl && hasUpper {
+					fmt.Fprintf(os.Stderr, "key %q has uppercase, shouldn't", key)
+				}
 				for _, value := range values {
 					fmt.Printf("Config: %s=%s\n", key, value)
 					source.Lines = append(source.Lines, fmt.Sprintf("%s=%s", key, value))
```

I'm not totally sold on the above, though, since it doesn't _really_ prevent us from breaking things in the future, it just prints some extra bits out to stderr.

Something that we could do to go even further, would be to teach the Values constructor to take a `*testing.T` instance, and call `Fail()` on it, instead of this. What do you think?

Reply:
I don't think the print to stderr will be effective, since I think Go suppresses most output if the test is successful. Of course, if we use mixed case, the test probably won't be successful.

I'm a bit negative on passing test contexts into non-test code, personally. It will likely bring dependencies into the binary that we don't need, and it seems a bit unhygienic. However, if we know that this is always an error (and it certainly looks like that's the case), we could just call `panic(fmt.Sprintf("key %q has uppercase, shouldn't", key))`, which will certainly alert the developer making this mistake. What would you think about that?

Reply:
Ah, your approach is a good way forward. This suggestion came from a (now outdated) understanding that the `struct Values` was test-only code (which it used to be, but it's now part of the publicly exported surface area from this package). Of course, we could introduce a test-only variant, but I think that it's probably not necessary...